### PR TITLE
fix Android upload banner and make animation right

### PR DESCRIPTION
**Pull request by songgao**

### CODE REVIEW

File: shared/fs/footer/upload-container.js [hunk 1]
```diff
@@ -1,16 +1,42 @@
...
 import Upload, {type UploadProps} from './upload'
 import {formatDurationFromNowTo} from '../../util/timestamp'
 
 const mapStateToProps = (state: TypedState) => ({
   _uploads: state.fs.uploads,
 })
 
-const mergeProps = ({_uploads}, dispatchProps, ownProps) =>
+// NOTE flip this to show a button to debug the upload banner animations.
+const enableDebugUploadBanner = true
```
Comment (chrisnojima) [inline shared/fs/footer/upload-container.js]: 🔥

File: shared/fs/footer/upload-container.js [hunk 2]
```diff
@@ -1,16 +1,42 @@
...
+
+const mapDispatchToProps = (dispatch: Dispatch) => ({
+  debugToggleShow: !enableDebugUploadBanner
+    ? undefined
+    : (() => {
```
Comment (chrisnojima) [inline shared/fs/footer/upload-container.js]: lets not put this in here inline. just make a closure up top. and likely wrap this in a if (__DEV__)

File: shared/fs/footer/upload-container.js [hunk 3]
```diff
@@ -1,16 +1,42 @@
...
+          tmp = !tmp
+        }
+      })(),
+})
+
+const mergeProps = ({_uploads}, {debugToggleShow}) =>
   ({
     files: _uploads.syncingPaths.merge(_uploads.writingToJournal).size,
-    timeLeft: formatDurationFromNowTo(_uploads.endEstimate),
+    timeLeft: '32 s' || formatDurationFromNowTo(_uploads.endEstimate),
```
Comment (chrisnojima) [inline shared/fs/footer/upload-container.js]: i assume this is debug leftover? 🔥

File: shared/fs/footer/upload.native.js [hunk 4]
```diff
@@ -1,66 +1,145 @@
...
+      return
+    }
+    this.animations[animation].stop()
+    this.animations[animation] = null
+  }
+
+  enter() {
+    this.setState({mounted: true})
+    this.stopAnimation('out')
+    this.startAnimationIn()
```
Comment (chrisnojima) [inline shared/fs/footer/upload.native.js]: don't you need to stop in and loop also before starting them again?

File: shared/fs/footer/upload.native.js [hunk 5]
```diff
@@ -1,66 +1,145 @@
...
+    this.startAnimationLoop()
+  }
+
+  exit() {
+    this.stopAnimation('in')
+    this.startAnimationOut(() => {
+      this.stopAnimation('loop')
+      this.setState({mounted: false})
```
Comment (chrisnojima) [inline shared/fs/footer/upload.native.js]: i don't believe this is safe

Comment (songgao) [reply shared/fs/footer/upload.native.js]: Could you explain why this is not safe? It's in the callback that's called when the animation has finished cleanly.

Comment (chrisnojima) [reply shared/fs/footer/upload.native.js]: any asynchronous callback can't safely call setState as the component could be unmounted. you either need to bookkeep the mounting or cancel any calls like this while unmounting

Comment (songgao) [reply shared/fs/footer/upload.native.js]: I see. What's the consequences of calling it in async and race happens?

In this case, I need this to determine whether to render the banner or not. If the whole component is unmounted, it doesn't really matter. Also, the component is always mounted as long as user is in the Files tab.

By bookkeeping, do you mean I should have a class variable that's set to `true` in `componentWillUnmount`, and only call this `setState` if that variable is `false`?

Comment (chrisnojima) [reply shared/fs/footer/upload.native.js]: react will give you warnings telling you its unsafe to do this. this is just a general rule. so if you do
```
setTimeout(() => this.setState({foo: 'bar'}), 10000)
```
that's unsafe.

yes, by bookkeeping i mean either

```
this.timeoutID = setTimeout(() => this.setState({foo: 'bar'}), 10000)

componentWillUnmount() {
clearTimeout(this.timeoutID)
}

// or

_mounted = true
componentWillUnmount () {
this._mounted = false
}

componentDidMount () {
this._mounted = true
}

setTimeout(() => this._mounted && this.setState({foo: 'bar'}), 10000)
```

**Pull request merged**